### PR TITLE
Add HF's Accelerate

### PR DESCRIPTION
**Pull request by ell-hol**
Training is much faster now

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -75,32 +76,39 @@ def __len__(self):
...
-        print(f"%s \n\n %s", (prime, "*" * 100))
-
-        sample = model.generate(GENERATE_LENGTH, inp[None, ...])
-        output_str = decode_tokens(sample[0])
-        print(output_str, "\n")
+
+        accelerator.backward(loss)
+
+        print(f"training loss: {loss.item()}")
+        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.5)
```
Comment (lucidrains) [inline train.py:96]: looks good! i think for gradient clipping, you'll need to use their helper function too?

**Comment by lucidrains** [comment]
do you want to try outfitting the PPO training too?

**Comment by lucidrains** [comment]
@ell-hol actually don't worry about the PPO, i'll just do it, since it is a bit out of the ordinary

**Comment by ell-hol** [comment]
> @ell-hol thank Yacine!

My pleasure. What about that wandb integration ?

**Comment by tengerye** [comment]
Hi @lucidrains @ell-hol  sorry to bother, but I found the Accelerate does not take effect. I designed two experiments to support my claim.

In the **first experiment**, I made changes to the `main.py` as follows, while others remain the same:
```
palm = PaLM(
    num_tokens = 20000,
    dim = 512,
    depth = 120,
).cuda()

prompts = torch.randint(0, 256, (50000, 4096)).cuda()
```
after I executed `python main.py`,
it ran out of GPU memory and it only used one GPU.


In the second experiment, I ran with the command `accelerate launch main.py`, and got the following error:
```
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/main.py", line 38, in <module>
    trainer.train(num_episodes = 50000, update_timesteps = 5)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/ppo.py", line 615, in train
    ) = self.actor_critic.generate(
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'DistributedDataParallel' object has no attribute 'generate'
episodes:   0%|                                                                       | 0/50000 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/main.py", line 38, in <module>
    trainer.train(num_episodes = 50000, update_timesteps = 5)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/ppo.py", line 615, in train
    ) = self.actor_critic.generate(
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1695, in __getattr__
    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'DistributedDataParallel' object has no attribute 'generate'
```

In the third experiment, I changed the lines:
```
                (
                    actions,
                    sequence,
                    mask,
                    prompt_mask,
                    action_logits,
                    value
                ) = self.actor_critic.generate(
                    rearrange(state, 'n -> 1 n'),
                    max_seq_len = max_seq_len,
                    eos_token = eos_token,
                    temperature = temperature,
                    return_values = True
                )
```
to
```
                (
                    actions,
                    sequence,
                    mask,
                    prompt_mask,
                    action_logits,
                    value
                ) = self.accelerate.unwrap_model(self.actor_critic).generate(
                    rearrange(state, 'n -> 1 n'),
                    max_seq_len = max_seq_len,
                    eos_token = eos_token,
                    temperature = temperature,
                    return_values = True
                )
```,

but I got the following error:
```
Traceback (most recent call last):
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/main.py", line 38, in <module>
    trainer.train(num_episodes = 50000, update_timesteps = 5)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/ppo.py", line 615, in train
    ) = self.accelerate.unwrap_model(self.actor_critic).generate(
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/utils.py", line 17, in inner
    out = fn(self, *args, **kwargs)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/ppo.py", line 116, in generate
    actions = self.actor_palm.generate(
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/utils.py", line 17, in inner
    out = fn(self, *args, **kwargs)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/palm.py", line 430, in generate
    logits, embeds = self.forward(out, return_logits_with_embedding = True, **kwargs)
  File "/home/ubuntu/workspace/PaLM-rlhf-pytorch/palm_rlhf_pytorch/palm.py", line 480, in forward
    x = self.token_emb(x)
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/nn/modules/sparse.py", line 162, in forward
    return F.embedding(
  File "/opt/conda/envs/rag/lib/python3.10/site-packages/torch/nn/functional.py", line 2233, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:1 and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)
```

**Comment by ell-hol** [comment]
@tengerye For train.py, you can just replace line 115 with :

```python
sample = model.module.generate(GENERATE_LENGTH, inp[None, ...])
``` 

Similarly, lines 607-620 of ppo.py become :
```python
            (
                actions,
                sequence,
                mask,
                prompt_mask,
                action_logits,
                value
            ) = self.actor_critic.module.generate(
                rearrange(state, 'n -> 1 n'),
                max_seq_len = max_seq_len,
                eos_token = eos_token,
                temperature = temperature,
                return_values = True
            )
``` 
Moreover, you shouldn't use .cuda() when you're using accelerate because .cuda() automatically moves your models and data to a single GPU: GPU number 0.
Instead of .cuda(), use device=accelerator.device and .to(device) as well as accelerator.prepare(), accelerator.backward() and all the other accelerate utilities just like I do in train.py
This worked for me this morning both in pre-training of PaLM and in the RLHF training on a 3-GPUs setup.

PS: I will submit a pull request but I'm not sure @lucidrains is still mainting this repo

**Pull request merged**